ui/src/app/api.service.ts

```diff
@@ -4,107 +4,247 @@
+    destroyed: boolean = false;
+    msg: any;
+    apiErrorHandler?: APIErrorHandler;
 
-    constructor(call_id: number, api_name: string) {
-        super();
+    constructor(call_id: number, api_name: string, api_args: any, resultHandler: ResultHandler, apiErrorHandler?: APIErrorHandler) {
         this.call_id = call_id;
         this.api_name = api_name;
+        this.resultHandler = resultHandler;
+        this.apiErrorHandler = apiErrorHandler;
+        this.msg = encode({
+            call_id: call_id,
+            api_name: api_name,
+            api_args: api_args,
+        });
+    }
+
+    destroy() {
+        this.destroyed = true;
+    }
+
+    handleResult(api_result: any) {
+        this.resultHandler(api_result);
+        this.destroy();
     }
 }
 
-type ReplyMsg = {
-    api_error: string;
-    call_id: number;
-    api_result: any;
-};
+//TODO(robnagler) needs to be generated by server
+const AUTH_API_NAME = 'pykern_http_auth';
+
+const AUTH_API_VERSION = 1;
 
 
 @Injectable({
     providedIn: 'root'
 })
 export class APIService {
-    private _socket$: WebSocketSubject<any>;
-    private _pendingCalls: Map<number, any> = new Map<number,any>();
-    private _destroyed: boolean = false;
-    private _receiver: Subscription;
-    private _call_id: number = 0;
-
-    constructor() {
-        console.log("constructor");
-        this._socket$ = webSocket({
-            url: '/api-v1',
-            binaryType: 'arraybuffer',
-            serializer: (v) => v,
-            deserializer: (v) => v,
-        });
-        this._receiver = this._socket$.asObservable().subscribe({
-            error: (err) => this._error(err),
-            next: (msg) => this._reply(msg),
-            complete: () => this._complete(),
-        })
+    #authOK: boolean = false;
+    #call_id: number = 0;
+    #client_id: string | null = null;
+    //TODO(robnagler) registration interface
+    #errorHandler = this.#defaultErrorHandler;
+    #pendingCalls: Map<number, Call> = new Map<number,Call>();
+    #socket: WebSocket | null = null;
+    #socketRetryBackoff: number = 0;
+    #unsentCalls: Array<Call> = new Array<Call>;
+
+    constructor(private log: LogService) {
+        this.log.info('create websocket');
+        this.#socketOpen();
     }
 
     // Send a message to the server
-    call(api_name: string, api_args: any) : Observable<any> {
-        const c = new Call(++this._call_id, api_name);
-        this._socket$.next(encode({
-            call_id: c.call_id,
-            api_name: api_name,
-            api_args: api_args,
-        }));
-        this._pendingCalls.set(c.call_id, c);
-        console.log("call", c.call_id, api_name, api_args);
-        return c;
+    call(api_name: string, api_args: any, resultHandler: ResultHandler, apiErrorHandler?: APIErrorHandler) : Call {
+        const rv = new Call(++this.#call_id, api_name, api_args, resultHandler, apiErrorHandler);
+        this.#unsentCalls.push(rv);
+        this.#send();
```

Comment:
Do you prefer the # for private members rather than defining it with the private keyword? I think it junks up the name and makes it harder to read when it is used.